### PR TITLE
Feature/6 create password

### PR DESCRIPTION
**Pull request by Rebal67**

### CODE REVIEW

File: src/user.ts [hunk 1]
```diff
@@ -16,20 +18,37 @@ class User {
...
+        console.log(accountName, masterPassword);
     }
 
-    generatePrivateKeyFromPassword(password: string) {
+
+    async generatePrivateKeyFromPassword(password: string): Promise<{ privateKey: PrivateKey, salt: Buffer }> {
         // creates a key based on secure (hashing) key generation algorithm like Argon2 or Scrypt
+        const randomString = crypto.randomBytes(32).toString('hex')
+        const salt = Buffer.from(randomString)
+        const hash = await argon2.hash(password, { salt })
```
Comment (theblockstalk) [inline src/user.ts:29]: will argon2 work in a browser, and react native environment?

File: src/user.ts [hunk 2]
```diff
@@ -16,20 +18,37 @@ class User {
...
-
-        const passwordPublicKey = this.authenticator.storeKey(AuthenticatorLevel.Password, privateKey, masterPassword);
-
-        this.id.create(accountName, passwordPublicKey, salt);
+        console.log(accountName, masterPassword);
     }
 
-    generatePrivateKeyFromPassword(password: string) {
+
+    async generatePrivateKeyFromPassword(password: string): Promise<{ privateKey: PrivateKey, salt: Buffer }> {
```
Comment (theblockstalk) [inline src/user.ts:25]: why are there two functions? generatePrivateKeyFromPassword and generateRandomPrivateKey?

Comment (Rebal67) [reply src/user.ts:25]: one that generates privateKey from password
and the other generates a random Key using the same algorithm

Comment (Rebal67) [reply src/user.ts:25]: also both functions are in the design dooc https://www.notion.so/tonomy-foundation/2022-08-03-Tonomy-Create-Account-feature-Design-doc-9f270a2dedbd4991abbbac2226e16d07

File: test/user.test.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  test('function savePassword is defined', () => {
+
+    expect(user.savePassword).toBeDefined();
+  });
+
+  test('generate private key returns privatekey', async () => {
+
+    const { privateKey } = await user.generatePrivateKeyFromPassword('123')
+
+    expect(privateKey).toBeInstanceOf(PrivateKey);
```
Comment (theblockstalk) [inline test/user.test.ts:21]: add test that salt exists as well

File: test/user.test.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+  test('password can be verfied', async () => {
+    const password = '123'
+    const { privateKey, salt } = await user.generatePrivateKeyFromPassword(password);
+    const data = Buffer.from(privateKey.data.array).toString('utf-8');
+    const result = await argon2.verify(data, password, { salt })
+    expect(result).toBe(true);
+  })
+
+
```
Comment (theblockstalk) [inline test/user.test.ts:33]: remove unused new lines. only need one space. in several places

File: test/user.test.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  test('function generateRandomPrivateKey is defined', () => {
+    expect(user.generateRandomPrivateKey).toBeDefined();
+  })
+
+  test('generate random key', async () => {
+    const r1 = user.generateRandomPrivateKey();
+    expect(r1).toBeInstanceOf(PrivateKey);
+
+    const r2 = user.generateRandomPrivateKey();
+    expect(r1).not.toBe(r2);
```
Comment (theblockstalk) [inline test/user.test.ts:51]: I'm not sure if `toBe()` will de a deep compare on the object, so may not be sufficient.

better to compare r1.toString(), r2.toString() to be sure

File: src/user.ts [hunk 6]
```diff
@@ -16,20 +18,38 @@ class User {
...
+        const { privateKey, salt } = await this.generatePrivateKeyFromPassword(masterPassword);
+        this.salt = salt;
+        const level = AuthenticatorLevel.Password;
+        this.authenticator.storeKey({ level, privateKey, challenge: masterPassword });
+    }
+
+    // Creates a cryptographically secure Private key
+    async generateRandomPrivateKey(): Promise<{ privateKey: PrivateKey, salt: Buffer }> {
+        const randomString = crypto.randomBytes(32).toString('hex')
+        const { privateKey, salt } = await this.generatePrivateKeyFromPassword(randomString)
```
[the remark this answers is not in the document]
Comment (Rebal67) [reply src/user.ts]: you are right

File: src/user.ts [hunk 7]
```diff
@@ -16,20 +18,38 @@ class User {
...
+
+    async savePassword(masterPassword: string) {
+        const { privateKey, salt } = await this.generatePrivateKeyFromPassword(masterPassword);
+        this.salt = salt;
+        const level = AuthenticatorLevel.Password;
+        this.authenticator.storeKey({ level, privateKey, challenge: masterPassword });
+    }
+
+    // Creates a cryptographically secure Private key
+    async generateRandomPrivateKey(): Promise<{ privateKey: PrivateKey, salt: Buffer }> {
```
[the remark this answers is not in the document]
Comment (Rebal67) [reply src/user.ts]: done

File: src/user.ts [hunk 8]
```diff
@@ -16,20 +18,38 @@ class User {
...
+        const randomString = crypto.randomBytes(32).toString('hex')
```
[the remark this answers is not in the document]
Comment (Rebal67) [reply src/user.ts]: toString removed

File: test/user.test.ts (hunk 3 above)
Comment (Rebal67) [reply test/user.test.ts:20]: done

File: test/user.test.ts (hunk 5 above)
Comment (Rebal67) [reply test/user.test.ts]: you are right. Gor object comparing toEqual is the correct function

File: test/user.test.ts (hunk 4 above)
Comment (Rebal67) [reply test/user.test.ts]: removed un used lines

**Review by theblockstalk — APPROVED** [review]

**Pull request merged**